README.md
# pcron

[![NPM version][npm-image]][npm-url]
[![NPM downloads][downloads-image]][downloads-url]
[![Build status][travis-image]][travis-url]

pcron is a periodical scheduler expression tool inspired by [fcron](http://fcron.free.fr/)

## Example

```
PT1h 2020 * 21-22 * * *
```

Like the example above, it means "from 21 through 22 of every month in 2020, the scheduler will be executed every 1 hour".

For more example, see [test](./test/index.test.ts)

## Installation

```
yarn add pcron
```

or

```
npm i pcron
```

## Usage

```ts
import { parseExpression } from "pcron";
import dayjs from "dayjs";

const timestamp = dayjs("2020-09-23T00:00:00").unix();
// timestamp is a unix timestamp, a current timestamp will be used by default if you omit this parameter
const exp = parseExpression("PT1s 2020 9 23 0 0-1 0-1", timestamp);
const nextDate = exp.next(); // you will get a dayjs object represent 2020-09-23T00:00:01
const nextDate = exp.next(); // you will get a dayjs object represent 2020-09-23T00:01:00
const nextDate = exp.next(); // you will get a dayjs object represent 2020-09-23T00:01:01
const nextDate = exp.next(); // you will get null
```

## Changelog

[change log](./CHANGELOG)

## License

MIT

[npm-image]: https://img.shields.io/npm/v/pcron.svg?style=flat
[npm-url]: https://npmjs.org/package/pcron
[downloads-image]: https://img.shields.io/npm/dm/pcron.svg?style=flat
[downloads-url]: https://npmjs.org/package/pcron
[travis-image]: https://img.shields.io/travis/duxiaofeng-github/pcron.svg?style=flat
[travis-url]: https://travis-ci.org/duxiaofeng-github/pcron


CHANGELOG
# 0.0.1

initial release

# 0.0.2

add week of month and week of year support

# 0.0.3

fix default export

# 0.0.4

fix default export

# 0.0.5

rebuild dist

# 0.0.6

fix sensitive case

test/index.test.ts
import { parseExpression } from "..";
import dayjs from "dayjs";
import zhCN from "dayjs/locale/zh-cn";

dayjs.locale(zhCN); // set the first day of week to Monday

// const timestamp = dayjs("2020-09-23T00:00:00").unix();

// test("parse expression", () => {
//   const exp = parseExpression("PT1s 2020 9 23 0 0-1 0-1", timestamp);
//   expect(exp.next().format()).toBe(dayjs("2020-09-23T00:00:01").format());
//   expect(exp.next().format()).toBe(dayjs("2020-09-23T00:01:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-09-23T00:01:01").format());
//   expect(exp.next()).toBeNull();
// });

// test("complete period expression", () => {
//   const exp = parseExpression("P1Y2M3DT4H5M6S * * * * * *", dayjs("2020-01-01T00:00:00").unix());
//   expect(exp.next().format()).toBe(dayjs("2021-03-04T04:05:06").format());
// });

test("minute period expression", () => {
  const exp = parseExpression("pt1m * * * * * *", dayjs("2020-01-01T00:00:00").unix());
  expect(exp.next().format()).toBe(dayjs("2020-01-01T00:01:00").format());
  expect(exp.next().format()).toBe(dayjs("2020-01-01T00:02:00").format());
  expect(exp.prev().format()).toBe(dayjs("2020-01-01T00:01:00").format());
});

// test("uncontinuous period expression", () => {
//   const exp = parseExpression("P1Y2DT3H4S * * * * * *", dayjs("2020-01-01T00:00:00").unix());
//   expect(exp.next().format()).toBe(dayjs("2021-01-03T03:00:04").format());
//   expect(exp.next().format()).toBe(dayjs("2022-01-05T06:00:08").format());
// });

// test("expression with multi-ranges", () => {
//   const exp = parseExpression("PT4s * * * * * 1-10,20-30", dayjs("2020-01-01T00:00:00").unix());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:01").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:05").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:09").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:20").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:24").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:28").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:01:01").format());
// });

// test("last one timestamp", () => {
//   const exp = parseExpression("PT4s 2020 1 1 0 0 1-10,20-30", dayjs("2020-01-01T00:00:00").unix());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:01").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:05").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:09").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:20").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:24").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:28").format());
//   expect(exp.next()).toBeNull();
//   expect(exp.prev().format()).toBe(dayjs("2020-01-01T00:00:28").format());
//   expect(exp.prev().format()).toBe(dayjs("2020-01-01T00:00:24").format());
//   expect(exp.prev().format()).toBe(dayjs("2020-01-01T00:00:20").format());
//   expect(exp.prev().format()).toBe(dayjs("2020-01-01T00:00:09").format());
//   expect(exp.prev().format()).toBe(dayjs("2020-01-01T00:00:05").format());
//   expect(exp.prev().format()).toBe(dayjs("2020-01-01T00:00:01").format());
//   expect(exp.prev()).toBeNull();
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:01").format());
// });

// test("continuous range expression", () => {
//   const exp = parseExpression("PT9s 2020 1 1 0 0 0-9,9-19,19-29,29-39,39-49,49-59", dayjs("2020-01-01T00:00:00").unix());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:09").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:18").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:27").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:36").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:45").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:54").format());
//   expect(exp.next()).toBeNull();
// });

// test("expression with week of month", () => {
//   const exp = parseExpression("P3D 2020 1-2 1-5/0-4 * * *", dayjs("2019-12-29T00:00:00").unix());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-06T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-09T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-13T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-16T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-20T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-23T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-27T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-30T00:00:00").format());

//   expect(exp.next().format()).toBe(dayjs("2020-02-03T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-02-06T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-02-10T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-02-13T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-02-17T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-02-20T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-02-24T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-02-27T00:00:00").format());
//   expect(exp.next()).toBeNull();

//   const exp2 = parseExpression("P1D * 5-6 1-6/5-6 * * *", dayjs("2020-04-30T00:00:00").unix());

//   // May of 2020 has 5 weekends and 6 weeks, June of 2020 has 4 weekends and 5 weeks
//   expect(exp2.next().format()).toBe(dayjs("2020-05-02T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-05-03T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-05-09T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-05-10T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-05-16T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-05-17T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-05-23T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-05-24T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-05-30T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-05-31T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-06-06T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-06-07T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-06-13T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-06-14T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-06-20T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-06-21T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-06-27T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2020-06-28T00:00:00").format());
//   expect(exp2.next().format()).toBe(dayjs("2021-05-01T00:00:00").format());
// });

// test("expression with week of year", () => {
//   const exp = parseExpression("P3D 2020 1-5/0-4 * * *", dayjs("2019-12-31T00:00:00").unix());
//   expect(exp.next().format()).toBe(dayjs("2020-01-01T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-06T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-09T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-13T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-16T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-20T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-23T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-27T00:00:00").format());
//   expect(exp.next().format()).toBe(dayjs("2020-01-30T00:00:00").format());

//   expect(exp.next()).toBeNull();
// });

// test("timestamp is out of range: next time should be next year", () => {
//   const exp = parseExpression("PT1s 2021-2022 * * * * *", timestamp);
//   expect(exp.next().format()).toBe(dayjs("2021-01-01T00:00:00").format());
// });

// test("timestamp is out of range: next time should be null", () => {
//   const exp = parseExpression("PT1s 2018-2019 * * * * *", timestamp);
//   expect(exp.next()).toBeNull();
// });

// test("timestamp is out of range: next time should be start of second range", () => {
//   const exp = parseExpression("PT1s 2018-2019,2021-2022 * * * * *", timestamp);
//   expect(exp.next().format()).toBe(dayjs("2021-01-01T00:00:00").format());
// });

// test("timestamp is out of range: next time after carrying of year should be null", () => {
//   const exp = parseExpression("PT1s 2020 6-7 * * * *", timestamp);
//   expect(exp.next()).toBeNull();
// });

// test("timestamp is out of range: next time after carrying of day should be Oct.", () => {
//   const exp = parseExpression("PT1s 2020 * 21-22 * * *", timestamp);
//   expect(exp.next().format()).toBe(dayjs("2020-10-21T00:00:00").format());
// });


src/index.ts
import dayjs, { Dayjs } from "dayjs";
import weekOfYear from "dayjs/plugin/weekOfYear";
import advancedFormat from "dayjs/plugin/advancedFormat";
import weekday from "dayjs/plugin/weekday";
import {
  Range,
  Options,
  Unit,
  getPrevUnit,
  getNextUnit,
  getPeriodByUnit,
  validateRange,
  transformUnitToDayjsUnit,
  getRangesByUnit,
  getValueByUnit,
  withDefaultRanges,
  getRangeInfo,
  isWeek,
  manipulateTimeByUnit,
  isWeekOfYear,
} from "./utils";

dayjs.extend(weekday);
dayjs.extend(weekOfYear);
dayjs.extend(advancedFormat);

function parseRange(range: string, type: Unit): Range[] {
  if (range === "*") {
    return [];
  }

  const rangeArray = range.split(",");

  return rangeArray
    .map((item) => {
      const [startString, endString] = item.split("-");
      const start = parseInt(startString);
      const end = endString != null ? parseInt(endString) : start;

      if (!validateRange(start, end, type)) {
        throw new Error(`Invalid range ${start}-${end} in`);
      }

      return { start, end };
    })
    .sort((a, b) => a.start - b.start);
}

function parseBlocks(options: { period: string; year: string; month?: string; weekOfYear?: string; weekOfMonth?: string; weekday?: string; day?: string; hour: string; min: string; sec: string }) {
  const { period, year, month, weekOfYear, weekOfMonth, weekday, day, hour, min, sec } = options;
  const parsedYear = parseRange(year, Unit.Year);
  const parsedMonth = month != null ? parseRange(month, Unit.Month) : undefined;
  const parsedWeekOfYear = weekOfYear != null ? parseRange(weekOfYear, Unit.WeekOfYear) : undefined;
  const parsedWeekOfMonth = weekOfMonth != null ? parseRange(weekOfMonth, Unit.WeekOfMonth) : undefined;
  const parsedWeekday = weekday != null ? parseRange(weekday, Unit.Weekday) : undefined;
  const parsedDay = day != null ? parseRange(day, Unit.Day) : undefined;
  const parsedHour = parseRange(hour, Unit.Hour);
  const parsedMin = parseRange(min, Unit.Min);
  const parsedSec = parseRange(sec, Unit.Sec);

  return {
    period,
    year: parsedYear,
    month: parsedMonth,
    weekOfYear: parsedWeekOfYear,
    weekOfMonth: parsedWeekOfMonth,
    weekday: parsedWeekday,
    day: parsedDay,
    hour: parsedHour,
    min: parsedMin,
    sec: parsedSec,
  };
}

function addOrSubtractDuration(time: Dayjs, duration: number, unit: Unit, isNext: boolean) {
  const opUnit = transformUnitToDayjsUnit(unit);

  return isNext ? time.add(duration, opUnit) : time.subtract(duration, opUnit);
}

function getEdgeOfRange(value: number, range: Range, periodNumber: number, isNext: boolean) {
  if (!isNext && periodNumber === 0) {
    return value;
  }

  const edge = isNext ? range.start : range.end - ((range.end - range.start) % periodNumber);

  return edge;
}

function getDurationByPeriodAndValue(value: number, periodNumber: number, range: Range, isNext: boolean) {
  if (isNext) {
    return periodNumber;
  } else {
    if (periodNumber === 0) {
      return periodNumber;
    } else {
      const mod = (value - range.start) % periodNumber;

      return mod === 0 ? periodNumber : mod;
    }
  }
}

function moveToNextRangeByUnit(unit: Unit, time: Dayjs, ranges: Range[], rangeIndex: number, options: Options, isNext: boolean) {
  let nextRangeIndex = isNext ? rangeIndex + 1 : rangeIndex - 1;
  let needToCarry = false;

  if (nextRangeIndex > ranges.length - 1) {
    nextRangeIndex = 0;
    needToCarry = true;
  } else if (nextRangeIndex < 0) {
    nextRangeIndex = ranges.length - 1;
    needToCarry = true;
  }

  let newTime = time;
  const prevUnit = getPrevUnit(unit, options);

  if (needToCarry) {
    if (prevUnit == null) {
      return null;
    } else {
      const prevUnitValue = getValueByUnit(newTime, prevUnit);
      newTime = manipulateTimeByUnit(newTime, prevUnit, prevUnitValue + 1);
    }
  }

  const { period } = options;
  const periodNumber = getPeriodByUnit(period, unit);
  const value = getValueByUnit(newTime, unit);
  const newRanges = withDefaultRanges(unit, getRangesByUnit(unit, newTime, options));

  if (newRanges != null) {
    const nextRange = newRanges[nextRangeIndex];
    const currentUnitEdge = getEdgeOfRange(value, nextRange, periodNumber, isNext);

    newTime = manipulateTimeByUnit(newTime, unit, currentUnitEdge);
  }

  const nextUnits = [];
  let nextUnit: Unit | null = unit;

  while (nextUnit != null) {
    nextUnit = getNextUnit(nextUnit, options);

    if (nextUnit != null) {
      nextUnits.push(nextUnit);
    }
  }

  nextUnits.forEach((nextUnit) => {
    const nextUnitRanges = withDefaultRanges(nextUnit, getRangesByUnit(nextUnit, newTime, options));

    if (nextUnitRanges != null) {
      const nextUnitRange = isNext ? nextUnitRanges[0] : nextUnitRanges[nextUnitRanges.length - 1];
      const nextUnitPeriodNumber = getPeriodByUnit(period, nextUnit);

      const nextUnitValue = getValueByUnit(newTime, nextUnit);
      const nextUnitEdge = getEdgeOfRange(nextUnitValue, nextUnitRange, nextUnitPeriodNumber, isNext);

      newTime = manipulateTimeByUnit(newTime, nextUnit, nextUnitEdge);
    }
  });

  return newTime;
}

function moveIntoRangeByUnit(unit: Unit, time: Dayjs, options: Options, isNext: boolean): Dayjs | null {
  const value = getValueByUnit(time, unit);
  const ranges = withDefaultRanges(unit, getRangesByUnit(unit, time, options));

  if (ranges) {
    if (ranges.length) {
      const { rangeIndex, inRange } = getRangeInfo(value, ranges, isNext);

      if (!inRange) {
        return moveToNextRangeByUnit(unit, time, ranges, rangeIndex, options, isNext);
      }
    }
  }

  return time;
}

function moveIntoRangeRecursively(unit: Unit, time: Dayjs, options: Options, isNext: boolean): Dayjs | null {
  const newTime = moveIntoRangeByUnit(unit, time, options, isNext);

  if (newTime == null) {
    return null;
  }

  const prevUnit = getPrevUnit(unit, options);

  if (prevUnit) {
    return moveIntoRangeRecursively(prevUnit, newTime, options, isNext);
  } else {
    return newTime;
  }
}

function moveToNextPeriodRecursively(unit: Unit, time: Dayjs, options: Options, isNext: boolean): Dayjs {
  const { period } = options;
  const periodNumber = getPeriodByUnit(period, unit);
  const value = getValueByUnit(time, unit);
  const ranges = withDefaultRanges(unit, getRangesByUnit(unit, time, options));

  let newTime = time;

  if (ranges) {
    if (ranges.length) {
      const { rangeIndex } = getRangeInfo(value, ranges, isNext);
      const range = ranges[rangeIndex];
      const duration = getDurationByPeriodAndValue(value, periodNumber, range, isNext);

      newTime = addOrSubtractDuration(newTime, duration, unit, isNext);
    }
  }

  const prevUnit = getPrevUnit(unit, options);

  if (prevUnit) {
    return moveToNextPeriodRecursively(prevUnit, newTime, options, isNext);
  } else {
    return newTime;
  }
}

function moveToNextPeriod(time: Dayjs, options: Options, isNext: boolean) {
  const timeMustBeInRange = moveIntoRangeRecursively(Unit.Sec, time, options, isNext);

  if (timeMustBeInRange == null || !time.isSame(timeMustBeInRange)) {
    return timeMustBeInRange;
  }

  const timeAfterPeriodIncreased = moveToNextPeriodRecursively(Unit.Sec, timeMustBeInRange, options, isNext);

  return moveIntoRangeRecursively(Unit.Sec, timeAfterPeriodIncreased, options, isNext);
}

class Result {
  private options: Options;
  private originalTimestamp: number;
  private currentTime: Dayjs;
  private hasPrev = true;
  private hasNext = true;

  constructor(options: Options, timestamp = dayjs().unix()) {
    this.options = options;
    this.originalTimestamp = timestamp;
    this.currentTime = dayjs.unix(timestamp);
  }

  reset() {
    this.currentTime = dayjs.unix(this.originalTimestamp);
  }

  prev() {
    if (!this.hasPrev) {
      return null;
    }

    if (!this.hasNext) {
      this.hasNext = true;

      return this.currentTime;
    }

    const newTime = moveToNextPeriod(this.currentTime, this.options, false);

    this.hasPrev = newTime != null;

    if (newTime != null) {
      this.currentTime = newTime;
    }

    return newTime;
  }

  next() {
    if (!this.hasNext) {
      return null;
    }

    if (!this.hasPrev) {
      this.hasPrev = true;

      return this.currentTime;
    }

    const newTime = moveToNextPeriod(this.currentTime, this.options, true);

    this.hasNext = newTime != null;

    if (newTime != null) {
      this.currentTime = newTime;
    }

    return newTime;
  }
}

export function parseExpression(expression: string, timestamp?: number) {
  const blocks = expression.split(" ");
  let options: Options;

  if (isWeek(expression)) {
    if (isWeekOfYear(blocks)) {
      const [period, year, weekOfYear, hour, min, sec] = blocks;
      const [week, weekday] = weekOfYear.split("/");

      options = parseBlocks({ period, year, weekOfYear: week, weekday, hour, min, sec });
    } else {
      const [period, year, month, weekOfMonth, hour, min, sec] = blocks;
      const [week, weekday] = weekOfMonth.split("/");

      options = parseBlocks({ period, year, month, weekOfMonth: week, weekday, hour, min, sec });
    }
  } else {
    const [period, year, month, day, hour, min, sec] = expression.split(" ");

    options = parseBlocks({ period, year, month, day, hour, min, sec });
  }

  return new Result(options, timestamp);
}

export default {
  parseExpression,
};


src/utils.ts
import dayjs, { Dayjs, OpUnitType } from "dayjs";

export interface Range {
  start: number;
  end: number;
}

export interface Options {
  period: string;
  year: Range[];
  month?: Range[];
  weekOfYear?: Range[];
  weekOfMonth?: Range[];
  weekday?: Range[];
  day?: Range[];
  hour: Range[];
  min: Range[];
  sec: Range[];
}

export enum Unit {
  Year = "year",
  Month = "month",
  Day = "day",
  WeekOfYear = "weekOfYear",
  Weekday = "weekday",
  WeekOfMonth = "weekOfMonth",
  Hour = "hour",
  Min = "minute",
  Sec = "second",
}

const defaultRanges = {
  [Unit.Year]: [-10000, 10000],
  [Unit.Month]: [1, 12],
  [Unit.Day]: [1, 31],
  [Unit.WeekOfYear]: [1, 53],
  [Unit.WeekOfMonth]: [1, 6],
  [Unit.Weekday]: [0, 6],
  [Unit.Hour]: [0, 23],
  [Unit.Min]: [0, 59],
  [Unit.Sec]: [0, 59],
};

function getDefaultRange(unit: Unit): Range[] {
  const range = defaultRanges[unit];

  return [{ start: range[0], end: range[1] }];
}

export function isWeek(expression: string) {
  return expression.indexOf("/") !== -1;
}

export function isWeekOfYear(blocks: string[]) {
  return blocks.length === 6;
}

export function getPrevUnit(unit: Unit, options: Options): Unit | null {
  const { weekOfMonth: weekOfMonthRanges, day: dayRanges } = options;

  switch (unit) {
    case Unit.Sec:
      return Unit.Min;
    case Unit.Min:
      return Unit.Hour;
    case Unit.Hour:
      return dayRanges != null ? Unit.Day : Unit.Weekday;
    case Unit.Day:
      return Unit.Month;
    case Unit.Weekday:
      return weekOfMonthRanges != null ? Unit.WeekOfMonth : Unit.WeekOfYear;
    case Unit.WeekOfMonth:
      return Unit.Month;
    case Unit.WeekOfYear:
      return Unit.Year;
    case Unit.Month:
      return Unit.Year;
    case Unit.Year:
      return null;
  }
}

export function getNextUnit(unit: Unit, options: Options): Unit | null {
  const { weekOfYear: weekOfYearRanges, weekOfMonth: weekOfMonthRanges } = options;

  switch (unit) {
    case Unit.Sec:
      return null;
    case Unit.Min:
      return Unit.Sec;
    case Unit.Hour:
      return Unit.Min;
    case Unit.Day:
      return Unit.Hour;
    case Unit.Weekday:
      return Unit.Hour;
    case Unit.WeekOfMonth:
      return Unit.Weekday;
    case Unit.WeekOfYear:
      return Unit.Weekday;
    case Unit.Month:
      return weekOfMonthRanges != null ? Unit.WeekOfMonth : Unit.Day;
    case Unit.Year:
      return weekOfYearRanges != null ? Unit.WeekOfYear : Unit.Month;
  }
}

export function getPeriodByUnit(period: string, unit: Unit): number {
  const [dateSection, timeSection = ""] = period.toLowerCase().replace(/p/i, "").split("t");
  let result;

  switch (unit) {
    case Unit.Sec:
      result = timeSection.match(/(\d+)s/i);
      break;
    case Unit.Min:
      result = timeSection.match(/(\d+)m/i);
      break;
    case Unit.Hour:
      result = timeSection.match(/(\d+)h/i);
      break;
    case Unit.Day:
    case Unit.Weekday:
      result = dateSection.match(/(\d+)d/i);
      break;
    case Unit.WeekOfMonth:
    case Unit.WeekOfYear:
      result = dateSection.match(/(\d+)w/i);
      break;
    case Unit.Month:
      result = dateSection.match(/(\d+)m/i);
      break;
    case Unit.Year:
      result = dateSection.match(/(\d+)y/i);
      break;
  }

  const periodNumber = result ? parseInt(result[1]) : 0;

  if (isNaN(periodNumber)) {
    throw new Error(`invalid period ${period}, unit: ${unit}`);
  }

  return periodNumber;
}

export function validateRange(start: number, end: number, type: Unit) {
  if (isNaN(start) || isNaN(end)) {
    return false;
  }

  const range = defaultRanges[type];

  if (range == null) {
    return true;
  }

  const [startEdge, endEdge] = range;

  return start >= startEdge && end <= endEdge;
}

export function transformUnitToDayjsUnit(unit: Unit): OpUnitType {
  switch (unit) {
    case Unit.Sec:
      return "second";
    case Unit.Min:
      return "minute";
    case Unit.Hour:
      return "hour";
    case Unit.Weekday:
      return "day";
    case Unit.WeekOfMonth:
      return "week";
    case Unit.WeekOfYear:
      return "week";
    case Unit.Day:
      return "day";
    case Unit.Month:
      return "month";
    case Unit.Year:
      return "year";
  }
}

export function getRangesByUnit(unit: Unit, time: Dayjs, options: Options): Range[] | undefined {
  const { year, month, weekOfYear, weekOfMonth, weekday, day, hour, min, sec } = options;

  switch (unit) {
    case Unit.Sec:
      return sec;
    case Unit.Min:
      return min;
    case Unit.Hour:
      return hour;
    case Unit.Day:
      return day;
    case Unit.Weekday:
      return weekday
        ? weekday.map((range) => {
            return adjustWeekdayRange(range, time);
          })
        : undefined;
    case Unit.WeekOfMonth:
      return weekOfMonth;
    case Unit.WeekOfYear:
      return weekOfYear;
    case Unit.Month:
      return month;
    case Unit.Year:
      return year;
  }
}

function adjustWeekdayRange(range: Range, time: Dayjs): Range {
  const firstWeek = 1;
  const startOfMonth = time.startOf("month");
  const endOfMonth = time.endOf("month");
  const firstWeekdayOfMonth = startOfMonth.weekday();
  const lastWeekdayOfMonth = endOfMonth.weekday();
  const lastWeekOfMonth = getWeek(endOfMonth) - getWeek(startOfMonth) + 1;
  const endOfYear = time.endOf("year");
  const lastWeekOfYear = getWeek(endOfYear);
  const lastWeekdayOfYear = endOfYear.weekday();
  const weekOfMonthValue = getValueByUnit(time, Unit.WeekOfMonth);
  const weekOfYearValue = getValueByUnit(time, Unit.WeekOfYear);

  if (weekOfMonthValue === firstWeek && range.start <= firstWeekdayOfMonth && range.end >= firstWeekdayOfMonth) {
    return { start: firstWeekdayOfMonth, end: range.end };
  } else if (weekOfMonthValue === lastWeekOfMonth && range.start <= lastWeekdayOfMonth && range.end >= lastWeekdayOfMonth) {
    return { start: range.start, end: lastWeekdayOfMonth };
  } else if (weekOfYearValue === lastWeekOfYear && range.start <= lastWeekdayOfYear && range.end >= lastWeekdayOfYear) {
    return { start: range.start, end: lastWeekdayOfYear };
  }

  return range;
}

function getWeek(time: Dayjs) {
  return time.week() === 0 && time.month() !== 0 ? time.subtract(1, "week").week() + 1 : time.week();
}

export function getValueByUnit(time: Dayjs, unit: Unit) {
  switch (unit) {
    case Unit.Sec:
      return time.second();
    case Unit.Min:
      return time.minute();
    case Unit.Hour:
      return time.hour();
    case Unit.Day:
      return time.date();
    case Unit.Weekday:
      return time.weekday();
    case Unit.WeekOfMonth:
      return getWeek(time) - getWeek(time.startOf("M")) + 1;
    case Unit.WeekOfYear:
      return getWeek(time);
    case Unit.Month:
      return time.month() + 1;
    case Unit.Year:
      return time.year();
  }
}

export function withDefaultRanges(unit: Unit, ranges: Range[] | undefined) {
  if (ranges == null) {
    return undefined;
  }

  if (ranges.length === 0) {
    return getDefaultRange(unit);
  }

  return ranges;
}

export function getRangeInfo(num: number, ranges: Range[], isNext: boolean) {
  let rangeIndex = isNext ? -1 : ranges.length;
  let inRange = false;

  for (let i = 0; i < ranges.length; i++) {
    const nextRangeIndex = isNext ? rangeIndex + 1 : rangeIndex - 1;
    const nextRange = ranges[nextRangeIndex];

    if ((isNext && num >= nextRange.start) || (!isNext && num <= nextRange.end)) {
      rangeIndex = nextRangeIndex;
    }

    if (num >= nextRange.start && num <= nextRange.end) {
      inRange = true;

      break;
    }
  }

  return { rangeIndex, inRange };
}

export function manipulateTimeByUnit(time: Dayjs, unit: Unit, value: number): Dayjs {
  switch (unit) {
    case Unit.Sec:
      return time.second(value);
    case Unit.Min:
      return time.minute(value);
    case Unit.Hour:
      return time.hour(value);
    case Unit.Day:
      return time.date(value);
    case Unit.Weekday:
      return time.weekday(value);
    case Unit.WeekOfMonth:
      const currentWeekday = time.weekday();

      return time.week(getWeek(time.startOf("M")) + value - 1).weekday(currentWeekday);
    case Unit.WeekOfYear:
      return time.week(value);
    case Unit.Month:
      return time.month(value - 1);
    case Unit.Year:
      return time.year(value);
  }
}
